### PR TITLE
[air/output] Reset heartbeat time in Ray Train after result

### LINKED ISSUE #36870: [AIR output]  Train example: Heartbeat info is not correct when verbose=2
### What happened + What you expected to happen

I ran this torch fine-tuning resnet example: https://docs.ray.io/en/latest/train/examples/pytorch/pytorch_resnet_finetune.html

The snippet of the output is attached below. The heartbeat info showed up twice:
`Training is in PENDING status. Current time: 2023-06-27 10:24:28. Total running time: 0s`
and
`Training on iteration 3. Current time: 2023-06-27 10:24:58. Total running time: 30s`

Based on our previous discussion: the rule is that we print per-iteration result when available and heartbeat info every 30s since the last update (result or heartbeat). 
- In this test, the heartbeat info got printed every 30s regardless of whether there is any result update or whether 30s passed since the last update. 
- The two pieces of heartbeat info above shouldn't get printed at all.

```
Training is in PENDING status. Current time: 2023-06-27 10:24:28. Total running time: 0s
Training started with configuration:
╭──────────────────────────────────────╮
│ Training config                      │
├──────────────────────────────────────┤
│ train_loop_config/batch_size      32 │
│ train_loop_config/input_size     224 │
│ train_loop_config/lr           0.001 │
│ train_loop_config/momentum       0.9 │
│ train_loop_config/num_epochs      10 │
╰──────────────────────────────────────╯

(TorchTrainer pid=5301, ip=10.0.0.86) Starting distributed worker processes: ['5353 (10.0.0.86)', '4778 (10.0.0.84)', '4961 (10.0.0.85)', '4801 (10.0.0.87)']
(RayTrainWorker pid=5353, ip=10.0.0.86) Setting up process group for: env:// [rank=0, world_size=4]
(RayTrainWorker pid=4961, ip=10.0.0.85) Moving model to device: cuda:0
(RayTrainWorker pid=4961, ip=10.0.0.85) Wrapping provided model in DistributedDataParallel.
Training finished iteration 1 at 2023-06-27 10:24:47. Total running time: 19s
╭───────────────────────────────────────────────╮
│ Training result                               │
├───────────────────────────────────────────────┤
│ time_this_iter_s                      13.5111 │
│ time_total_s                          13.5111 │
│ training_iteration                          1 │
│ acc                                   0.68421 │
│ config/train_loop_config/batch_size        32 │
│ config/train_loop_config/input_size       224 │
│ config/train_loop_config/lr             0.001 │
│ config/train_loop_config/momentum         0.9 │
│ config/train_loop_config/num_epochs        10 │
│ loss                                   0.5979 │
╰───────────────────────────────────────────────╯

Training saved a checkpoint for iteration 1 at: /tmp/ray_results/finetune-resnet/TorchTrainer_7800d_00000_0_2023-06-27_10-24-28/checkpoint_000000

Training finished iteration 2 at 2023-06-27 10:24:55. Total running time: 27s
╭───────────────────────────────────────────────╮
│ Training result                               │
├───────────────────────────────────────────────┤
│ time_this_iter_s                      8.25429 │
│ time_total_s                          21.7654 │
│ training_iteration                          2 │
│ acc                                   0.89474 │
│ config/train_loop_config/batch_size        32 │
│ config/train_loop_config/input_size       224 │
│ config/train_loop_config/lr             0.001 │
│ config/train_loop_config/momentum         0.9 │
│ config/train_loop_config/num_epochs        10 │
│ loss                                  0.40504 │
╰───────────────────────────────────────────────╯

Training saved a checkpoint for iteration 2 at: /tmp/ray_results/finetune-resnet/TorchTrainer_7800d_00000_0_2023-06-27_10-24-28/checkpoint_000001

Training on iteration 3. Current time: 2023-06-27 10:24:58. Total running time: 30s
Training finished iteration 3 at 2023-06-27 10:25:00. Total running time: 31s
╭───────────────────────────────────────────────╮
│ Training result                               │
├───────────────────────────────────────────────┤
│ time_this_iter_s                      4.26406 │
│ time_total_s                          26.0294 │
│ training_iteration                          3 │
│ acc                                   0.94737 │
│ config/train_loop_config/batch_size        32 │
│ config/train_loop_config/input_size       224 │
│ config/train_loop_config/lr             0.001 │
│ config/train_loop_config/momentum         0.9 │
│ config/train_loop_config/num_epochs        10 │
│ loss                                  0.30175 │
╰───────────────────────────────────────────────╯

Training saved a checkpoint for iteration 3 at: /tmp/ray_results/finetune-resnet/TorchTrainer_7800d_00000_0_2023-06-27_10-24-28/checkpoint_000002

Training finished iteration 4 at 2023-06-27 10:25:04. Total running time: 36s
╭───────────────────────────────────────────────╮
│ Training result                               │
├───────────────────────────────────────────────┤
│ time_this_iter_s                      4.37041 │
│ time_total_s                          30.3998 │
│ training_iteration                          4 │
│ acc                                   0.94737 │
│ config/train_loop_config/batch_size        32 │
│ config/train_loop_config/input_size       224 │
│ config/train_loop_config/lr             0.001 │
│ config/train_loop_config/momentum         0.9 │
│ config/train_loop_config/num_epochs        10 │
│ loss                                  0.23877 │
╰───────────────────────────────────────────────╯
```

### Versions / Dependencies

nightly

### Reproduction script

https://docs.ray.io/en/latest/train/examples/pytorch/pytorch_resnet_finetune.html

### Issue Severity

Low: It annoys or frustrates me.

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Includes #37531.

Currently Ray Train prints a heartbeat every 30 seconds. However, it should only print a heartbeat if it didn't receive a result in the meantime. 

With this PR, we reset the heartbeat time when a result arrives. It includes a test to assert the difference between Ray Train and Ray Tune in handling this case.

## Related issue number

Closes #36870

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/experimental/output.py [hunk 1]
```diff
@@ -1089,3 +1090,16 @@ def _get_heartbeat(self, trials: List[Trial], force_full_output: bool = False):
...
+
+    def on_trial_result(
+        self,
+        iteration: int,
+        trials: List[Trial],
+        trial: Trial,
+        result: Dict,
+        **info,
+    ):
+        self._last_heartbeat_time = time.time()
```
Comment (woshiyyya) [inline python/ray/tune/experimental/output.py:1102]: Why do we reset the heartbeat time here? Is it because we already print a metric table on trial result?

**Review by woshiyyya — APPROVED** [review]
LGTM. Thank you Kai!
